src/toolchain/names.rs

```diff
@@ -308,38 +308,31 @@ impl Display for ToolchainName {
     }
 }
 
+/// Sorts [`ToolchainName`]s in the following order:
```

Comment:
Sorry to be a PITA again, but would you mind splitting this into multiple commits to make it clearer how you got here? I think I see four here:

* Switch from `Version` to a tuple type for the key
* Switch from `sort_by()` to `sort_by_key()`
* Inlining `special_version()`
* Inlining `toolchain_sort_key()`

Reply:
@djc Please mind that the current semantics is not exactly the same as the previous panic-prone one (I worked backwards from the test case to figure out what I should do), so it has been a rewrite rather than a gradual refactoring, so in this case I think "faking" a refactor is not that effective. Of course, I'll be glad to do that if you wish so.

Reply:
I think that is exactly why it is important to split this, because I want to be able to have a smaller commit that only has the semantic changes so it's easier to review it in isolation. That doesn't seem like "faking" a refactor, but rather having the refactoring as a precursor to making the semantic changes more obvious.

Reply:
@djc I've split the history into 7 commits, and only [c27f007](https://github.com/rust-lang/rustup/pull/3518/commits/c27f007695d7776b1e88fd58c0056790682eab86) is an actual change.